sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala

```diff
@@ -60,6 +61,54 @@ class SQLOpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
 }
 
 object SQLOpenHashSet {
+  def withNullCheckFunc(
+      dataType: DataType,
+      hashSet: SQLOpenHashSet[Any],
+      handleNotNull: Any => Unit,
+      handleNull: () => Unit): (ArrayData, Int) => Unit = {
+    (array: ArrayData, index: Int) =>
+      if (array.isNullAt(index)) {
+        if (!hashSet.containsNull) {
+          hashSet.addNull
+          handleNull()
+        }
+      } else {
+        val elem = array.get(index, dataType)
+        handleNotNull(elem)
+      }
+  }
+
+  def withNullCheckCode(
+      arrayContainsNull: Boolean,
+      setContainsNull: Boolean,
+      hashSet: String,
+      handleNotNull: (String, String) => String,
+      handleNull: String): (String, String) => String = {
```

Comment:
instead of returning `(String, String) => String`, should this method take `array: String, index: String` and return `String` directly?

Reply:
> instead of returning `(String, String) => String`, should this method take `array: String, index: String` and return `String` directly?

For case ArrayUnion
```
        Seq(array1, array2).foreach { array =>
          var i = 0
          while (i < array.numElements()) {
            withNullCheckFunc(array, i)
            i += 1
          }
        }
```
We need to use array as a parameter